exercises/minesweeper/src/example.c

```diff
@@ -0,0 +1,66 @@
+#include "minesweeper.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <ctype.h>
+
+static void set_single_field(char **minefield, const size_t row,
+                             const size_t col)
+{
+   if (minefield[row][col] == '*')
+      return;
+
+   const char c = minefield[row][col];
+   minefield[row][col] = (c == ' ') ? '1' : c + 1;
+}
+
+static void set_fields(char **minefield, const size_t all_rows,
+                       const size_t all_cols, const size_t row,
+                       const size_t col)
+{
+   const size_t from_row = row ? row - 1 : row;
+   const size_t to_row = (row + 1 < all_rows) ? row + 1 : row;
+   const size_t from_col = col ? col - 1 : col;
+   const size_t to_col = (col + 1 < all_cols) ? col + 1 : col;
+
+   for (size_t i = from_row; i <= to_row; i++) {
+      for (size_t k = from_col; k <= to_col; k++)
+         set_single_field(minefield, i, k);
+   }
+}
+
+char **annotate(const char **minefield, const size_t rows)
+{
+   if (minefield == NULL || *minefield == NULL || rows == 0)
+      return NULL;
+
+   const size_t cols = strlen(minefield[0]);
+
+   char **annotated = malloc(sizeof(char *) * rows);
+   if (annotated == NULL) {
+      fprintf(stderr, "Memory error!\n");
+      return NULL;
+   }
+
+   for (size_t i = 0; i < rows; i++) {
+      annotated[i] = malloc(cols + 1);
+      if (annotated[i] == NULL) {
+         fprintf(stderr, "Memory error!\n");
+         free(annotated);
```

Comment:
Technically you have a memory leak here. If you have a `malloc` failure partway through here, you will only free the top level `annotated` array, but not all of the previous rows. One way to fix this without making your solution much more complicated is to allocate all of the rows with one `malloc` (all rows in one string) and then set the pointers from `annotated` for each row by indexing into the combined row string.

So the pointers in `annotated` would be set like this for a 3x3 mine field:

```
annotated = [ _ , _ , _ ]
                        |    \     \
                       /      |      \
                     /        |        \ 
all rows = [ |, _, _, |, _, _, |, _, _ ]
```

Reply:
Lol! You are completely right - I overlooked this.
Thanks for the suggestion, I like it - will take care of this today.